src/utils/allocators.zig

```diff
@@ -136,38 +136,78 @@ pub fn RecycleFBA(config: struct {
     };
 }
 
+test "recycle allocator" {
+    const backing_allocator = std.testing.allocator;
+    var allocator = try RecycleFBA(.{}).init(backing_allocator, 1024);
+    defer allocator.deinit();
+
+    // alloc a slice of 100 bytes
+    const bytes = try allocator.allocator().alloc(u8, 100);
+    const ptr = bytes.ptr;
+    // free the slice
+    allocator.allocator().free(bytes);
+
+    // realloc should be the same (ie, recycled data)
+    const bytes2 = try allocator.allocator().alloc(u8, 100);
+    try std.testing.expectEqual(ptr, bytes2.ptr);
+    allocator.allocator().free(bytes2);
+
+    // same result with smaller slice
+    const bytes3 = try allocator.allocator().alloc(u8, 50);
+    try std.testing.expectEqual(ptr, bytes3.ptr);
+    allocator.allocator().free(bytes3);
+
+    // diff result with larger slice
+    const bytes4 = try allocator.allocator().alloc(u8, 200);
+    try std.testing.expect(ptr != bytes4.ptr);
+    allocator.allocator().free(bytes4);
+}
+
 /// thread safe disk memory allocator
 pub const DiskMemoryAllocator = struct {
-    filepath: []const u8,
-    count: std.atomic.Value(u64) = std.atomic.Value(u64).init(0),
-
+    dir: std.fs.Dir,
+    logger: sig.trace.Logger,
+    /// The address space mmap'd to a particular file will be at least
+    /// `(file_size * (1000 + mmap_ratio)) / 1000` in integer terms.
+    /// With a value of 0, the mmap'd size will be equal to the file size.
+    /// With a value of 1000, the mmap'd size will be double the file size.
+    /// NOTE: not intended to be changed after initialization.
+    mmap_ratio: MmapRatio = 0,
+    count: std.atomic.Value(u32) = std.atomic.Value(u32).init(0),
     const Self = @This();
 
-    pub fn init(filepath: []const u8) Self {
-        return Self{
-            .filepath = filepath,
-        };
+    pub const MmapRatio = u32;
+
+    /// Metadata stored at the end of each allocation.
+    const Metadata = extern struct {
+        file_index: u32,
+        mmap_size: packed struct(u8) {
```

Comment:
Well, it saves 7 bytes which can instead be used for actual data in the user-facing allocation. The more memory used by the metadata, the sooner a second page is needed.

Also: the page size is always a power of two on all realistic platforms. I'm not sure there's a scenario where we are going to be running on a platform that's exotic enough for that to not be true.
You can technically request to mmap a size that's not aligned to the page size, but it will still just allocate a whole page - the kernel isn't in the business of tracking the size of your allocations at a granularity smaller than a page. So this is imo an entirely reasonable property/assumption to take advantage of.

The main reason this is a packed struct wrapping the integer is because the exotically sized integer isn't allowed in an extern struct. Would use a packed struct, but that makes it act bit-packed and backed by an integer, which has less intuitive properties on the size of how much space it takes up than using an extern struct.

Reply:
> Well, it saves 3 bytes which can instead be used for actual data in the user-facing allocation. The more memory used by the metadata, the sooner a second page is needed,

I understand this, it just seems negligible though. Pages are measured in thousands of bytes. The performance gain of this approach is less than 0.1%. It's hard to justify the code complexity. Actually this approach requires extra cpu operations to calculate logs/shifts, so it might even have a net negative impact on performance.

> the page size is always a power of two on all realistic platforms

That doesn't mean the mapped size is a power of two. Regardless of page size, 3 pages has 3 as a prime factor, which means it's not a power of two. To ensure the mapped size is a power of two, the number of pages per mapping would also need to be a power of two.

Reply:
Ah, a fair point, I was conflating my alignment math and power²/log₂ math.

Reply:
fc14c0746a8dca4c94fac28ab548c7a4795a3df0